### PR TITLE
[android] returning null instead of the Source object in the MapboxMap#removeSource when it fails

### LINKED ISSUE #12526: Removing source processed by the layer
When removing a `Source` that's being currently processed by a `Layer` we are failing silently and [printing a log](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/style/style_impl.cpp#L169-L172).

We should improve the experience during this unusual condition by maybe printing a higher level log, and on the Android side updating docs and returning `null` instead of the `Source` object in the `MapboxMap#removeSource` when it fails.

Example that will fail:
```
mapView.getMapAsync(mapboxMap -> {
  GeoJsonSource source = new GeoJsonSource("source", Point.fromLngLat(0, 0));
  mapboxMap.addSource(source);
  mapboxMap.addLayer(new CircleLayer("layer", "source"));
  mapboxMap.removeSource("source");
});
```

How's @mapbox/maps-ios handling this scenario?

### PR DESCRIPTION
**Pull request by osana**
closes #12526

### CODE REVIEW

**Comment by LukasPaczos** [comment]
I agree with @tobrun, `remove` methods might be more readable if they return boolean values indicating whether removal has been successful. Users can always fetch the layer/source before removal with `get` methods which should cover current use cases.

**Comment by osana** [comment]
@tobrun @LukasPaczos  Do you think it makes sense to change `removeLayer` return value 
in this PR?

**Comment by LukasPaczos** [comment]
Sure, shouldn't be an issue.

File: platform/android/src/native_map_view.cpp [hunk 1]
```diff
@@ -827,47 +827,49 @@ void NativeMapView::addLayerAt(JNIEnv& env, jlong nativeLayerPtr, jni::jint inde
...
+        return jni::jni_true;
     }
+    return jni::jni_false;
 }
 
 /**
  * Remove with wrapper object id. Ownership is transferred back to the wrapper
  */
-void NativeMapView::removeLayer(JNIEnv&, jlong layerPtr) {
+jni::jboolean NativeMapView::removeLayer(JNIEnv&, jlong layerPtr) {
```
Comment (LukasPaczos) [inline platform/android/src/native_map_view.cpp]: Every other `remove` method besides this one drops the reference to the layer, so the one kept in Java is probably going to become invalid. In order to make our intended logic possible, that a user fetches the layer with `getLayer` (or just keeps the reference from the start) and can keep after it's removed from the map (successfully or not), we need to keep the reference valid.

**Comment by osana** [comment]
@tobrun @LukasPaczos Capturing offline conversation of what needs to happen in this PR
1) `MapboxMap` has three `removeLayer` methods. At the moment two of them invalidate existing jav areferences to the removed layer. 
https://github.com/mapbox/mapbox-gl-native/blob/1ceea72e6da786537cf3f4038ce56b602c12bfa8/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java#L354
https://github.com/mapbox/mapbox-gl-native/blob/1ceea72e6da786537cf3f4038ce56b602c12bfa8/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java#L376

We need the above two methods keep the reference alive ( transfer ownership from native to java peer) as it is done in 
https://github.com/mapbox/mapbox-gl-native/blob/1ceea72e6da786537cf3f4038ce56b602c12bfa8/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java#L365

TODO: javadoc update & native_map_view.cpp change

2) Add test to verify  the above is working, ie the reference is still valid after the Layer was removed

3) MapboxMap has two methods to `removeSource`. They result in the same native method call.

File: platform/android/src/native_map_view.cpp [hunk 2]
```diff
@@ -827,47 +827,44 @@ void NativeMapView::addLayerAt(JNIEnv& env, jlong nativeLayerPtr, jni::jint inde
...
 
     std::unique_ptr<mbgl::style::Layer> coreLayer = map->getStyle().removeLayer(layers.at(index)->getID());
     if (coreLayer) {
-        return LayerManagerAndroid::get()->createJavaLayerPeer(env, *map, std::move(coreLayer));
-    } else {
-        return jni::Local<jni::Object<Layer>>();
+        jni::Local<jni::Object<Layer>> layerObj =
+                LayerManagerAndroid::get()->createJavaLayerPeer(env, *map, std::move(coreLayer));
+        mbgl::android::Layer *layer = reinterpret_cast<mbgl::android::Layer *>(layerObj.get());
+        layer->setLayer(std::move(coreLayer));
```
Comment (osana) [inline platform/android/src/native_map_view.cpp]: @tobrun could you ask someone with more knowledge of c++ how this should be done . Lines 846 -848 are not correct.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/NativeMapView.java [hunk 3]
```diff
@@ -748,29 +748,32 @@ public void addLayerAt(@NonNull Layer layer, @IntRange(from = 0) int index) {
...
 
-  @Nullable
-  public Layer removeLayerAt(@IntRange(from = 0) int index) {
+  public boolean removeLayerAt(@IntRange(from = 0) int index) {
     if (checkState("removeLayerAt")) {
-      return null;
+      return false;
     }
     return nativeRemoveLayerAt(index);
+
```
Comment (LukasPaczos) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/NativeMapView.java]: NIT: new line

**Pull request merged**